service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -0,0 +1,138 @@
+    private static final String SSP_INTERACTION_ID = "Ssp-InteractionID";
+    private static final int SDS_URI_OFFSET = 8;
+
+    private final SdsClient sdsClient;
+
+    private Map<String, Function<String, Optional<SdsClient.SdsResponseData>>> sdsRequestFunctions;
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (StringUtils.isBlank(System.getProperty(GPC_URL_ENVIRONMENT_VARIABLE))) {
+            ServerHttpRequest serverHttpRequest = exchange.getRequest();
+            extractInteractionId(serverHttpRequest.getHeaders())
+                .ifPresent(id -> proceedSdsLookup(serverHttpRequest, exchange, id));
+        }
+
+        return chain.filter(exchange);
+    }
+
+    @Override
+    public int getOrder() {
+        return RouteToRequestUrlFilter.ROUTE_TO_URL_FILTER_ORDER + 1;
+    }
+
+    @PostConstruct
+    @SuppressWarnings("unused")
+    public void initializeSdsRequestFunctions() {
+        sdsRequestFunctions = Map.of(
+            STRUCTURED_ID, sdsClient::callForGetStructuredRecord,
+            PATIENT_SEARCH_ID, sdsClient::callForPatientSearchAccessDocument,
+            DOCUMENT_SEARCH_ID, sdsClient::callForSearchForDocumentRecord,
+            BINARY_READ_ID, sdsClient::callForRetrieveDocumentRecord);
+    }
+
+    private void proceedSdsLookup(ServerHttpRequest serverHttpRequest,
+            ServerWebExchange exchange,
+            String integrationId) {
+        String organisation = extractOrganisation(serverHttpRequest.getPath());
+        SdsClient.SdsResponseData response = performRequestAccordingToInteractionId(integrationId, organisation)
+            .orElseThrow(() -> new SdsException(
+                String.format("No endpoint found in SDS for GP Connect endpoint InteractionId=%s OdsCode=%s",
+                    integrationId,
+                    organisation))
+            );
+        prepareLookupUri(response.getAddress(), serverHttpRequest.getPath())
+            .ifPresent(uri -> exchange.getAttributes()
+                .put(ServerWebExchangeUtils.GATEWAY_REQUEST_URL_ATTR, uri));
+    }
+
+    private Optional<SdsClient.SdsResponseData> performRequestAccordingToInteractionId(String interactionId,
+            String organisation) {
+        if (isRequestFunctionAvailable(interactionId)) {
+            LOGGER.info("Performing request with organisation \"{}\" and NHS service endpoint id \"{}\"",
+                organisation, interactionId);
+            return sdsRequestFunctions.get(interactionId)
+                .apply(organisation);
+        }
+        throw new IllegalArgumentException(String.format("Not recognised InteractionId %s", interactionId));
+    }
+
+    private boolean isRequestFunctionAvailable(String interactionId) {
+        return sdsRequestFunctions.keySet()
+            .stream()
+            .anyMatch(key -> key.equals(interactionId));
```

Comment:
Why iterate the keySet of a Map just to see if it contains a key?

```suggestion
        return sdsRequestFunctions.containsKey(interactionId);
```